impeller/renderer/backend/vulkan/command_pool_vk.h

```diff
@@ -5,109 +5,126 @@
 
-  /// @brief      Whether or not this |CommandPoolVK| is valid.
+  /// @brief      Creates a resource that manages the life of a command pool.
   ///
-  /// A command pool is no longer when valid once it's been |Reset|.
-  bool IsValid() const;
+  /// @param[in]  pool      The command pool to manage.
+  /// @param[in]  recycler  The context that will be notified on destruction.
+  explicit CommandPoolVK(vk::UniqueCommandPool pool,
+                         std::weak_ptr<ContextVK>& context)
+      : pool_(std::move(pool)), context_(context) {}
 
   /// @brief      Creates and returns a new |vk::CommandBuffer|.
   ///
-  /// An attempt is made to reuse existing buffers (instead of creating new
-  /// ones) by recycling buffers that have been collected by
-  /// |CollectGraphicsCommandBuffer|.
-  ///
   /// @return     Always returns a new |vk::CommandBuffer|, but if for any
   ///             reason a valid command buffer could not be created, it will be
   ///             a `{}` default instance (i.e. while being torn down).
-  vk::UniqueCommandBuffer CreateGraphicsCommandBuffer();
+  vk::UniqueCommandBuffer CreateBuffer();
 
-  /// @brief      Collects the given |vk::CommandBuffer| for recycling.
-  ///
-  /// The given |vk::CommandBuffer| will be recycled (reused) in the future when
-  /// |CreateGraphicsCommandBuffer| is called.
+  /// @brief      Collects the given |vk::CommandBuffer| to be retained.
   ///
   /// @param[in]  buffer  The |vk::CommandBuffer| to collect.
   ///
-  /// @note       This method is a noop if a different thread created the pool.
-  ///
   /// @see        |GarbageCollectBuffersIfAble|
-  void CollectGraphicsCommandBuffer(vk::UniqueCommandBuffer buffer);
+  void CollectBuffer(vk::UniqueCommandBuffer&& buffer);
 
  private:
-  const std::thread::id owner_id_;
-  std::weak_ptr<const DeviceHolder> device_holder_;
-  vk::UniqueCommandPool graphics_pool_;
-  Mutex buffers_to_collect_mutex_;
-  std::vector<vk::UniqueCommandBuffer> buffers_to_collect_
-      IPLR_GUARDED_BY(buffers_to_collect_mutex_);
-  std::vector<vk::UniqueCommandBuffer> recycled_buffers_;
-  bool is_valid_ = false;
-
-  /// @brief      Resets, releasing all |vk::CommandBuffer|s.
+  FML_DISALLOW_COPY_AND_ASSIGN(CommandPoolVK);
+
+  vk::UniqueCommandPool pool_;
+  std::weak_ptr<ContextVK>& context_;
+
+  // Used to retain a reference on these until the pool is reset.
+  std::vector<vk::UniqueCommandBuffer> collected_buffers_;
+};
+
+//------------------------------------------------------------------------------
+/// @brief      Creates and manages the lifecycle of |vk::CommandPool| objects.
+///
+/// A |vk::CommandPool| is expensive to create and reset. This class manages
+/// the lifecycle of |vk::CommandPool| objects by creating and recycling them;
+/// or in other words, a pool for command pools.
+///
+/// A single instance should be created per |ContextVK|.
+///
+/// Every "frame", a single |CommandPoolResourceVk| is made available for each
```

Comment:
Please make a note of the the threading restrictions you've accounted for here. Specifically, the spec says:
```
Command pools are externally synchronized, meaning that a command pool must not
be used concurrently in multiple threads. That includes use via recording
commands on any command buffers allocated from the pool, as well as
operations that allocate, free, and reset command buffers or the pool itself.
```

The things I would look for are:
1. Are the commands in the command pool done executing when recycle is called?
2. Does the swap and recycle happen in a thread safe manner.

Reply:
Ack. I did find a problem (in unit tests) and addressed them. Did you want the docs updated?

Reply:
Sounds good, added!